src/validation.cpp

```diff
@@ -4674,29 +4674,39 @@ bool LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFi
                 break;
             }
             try {
-                // read block
+                // read block header
                 uint64_t nBlockPos = blkdat.GetPos();
                 if (dbp)
                     dbp->nPos = nBlockPos;
-                blkdat.SetLimit(nBlockPos + nSize);
-                blkdat.SetPos(nBlockPos);
-                std::shared_ptr<CBlock> pblock = std::make_shared<CBlock>();
-                CBlock& block = *pblock;
-                blkdat >> block;
-                nRewind = blkdat.GetPos();
-
-                uint256 hash = block.GetHash();
+                blkdat.SetLimit(nBlockPos + 80);
+                CBlockHeader header;
+                blkdat >> header;
+
+                const uint256 hash = header.GetHash();
                 {
                     LOCK(cs_main);
                     // detect out of order blocks, and store them for later
-                    if (hash != chainparams.GetConsensus().hashGenesisBlock && !LookupBlockIndex(block.hashPrevBlock)) {
+                    if (hash != chainparams.GetConsensus().hashGenesisBlock && !LookupBlockIndex(header.hashPrevBlock)) {
                         LogPrint(BCLog::REINDEX, "%s: Out of order block %s, parent %s not known\n", __func__, hash.ToString(),
-                                block.hashPrevBlock.ToString());
+                                header.hashPrevBlock.ToString());
                         if (dbp)
-                            mapBlocksUnknownParent.insert(std::make_pair(block.hashPrevBlock, *dbp));
+                            mapBlocksUnknownParent.insert(std::make_pair(header.hashPrevBlock, *dbp));
+
+                        // Position to the start of the next block.
+                        nRewind = nBlockPos + nSize;
```

Comment:
I think this is incorrect. If the size field of the block was corrupted this would cause us to potentially jump into nowhere. Do things still work (and still give a performance benefit...) if you change this to nBlockPos + 1?

Reply:
I think if `nSize` as read from disk is wrong, then

(a) if it's smaller than the actual size of the block, `blkdat >> block` will currently throw which will spit out a `Deserialize or I/O error` and start trying again at the end of the `nSize` bytes -- so this behaviour should be near enough to the same, and this is fine

and (b) if it's larger than the actual size of the block, the `nSize` will be effectively ignored, and we'll pick up at a different, possibly incorrect point, and try to find the next `chainparams.MessageStart()` marker, so we could miss up to, I guess, 8MB worth of block data (4MB due to a bad nSize which is just skipped over, then just under 4MB if the next block was as large as possible and we just missed its start which is read byte-by-byte) due to 4 bytes of corrupted size data, rather than just coping with it. This is worse, but doesn't seem terrible -- if the four bytes that were corrupted were the last bytes of one block and the first bytes of another you'd have lost up to 8MB of data that way too, without any more precise targeting.

Reply:
The main point of `-reindex` now is recovery after on-disk corruption, so I think it makes sense to be as robust to that as we can reasonably be?

As well as just random disk corruption, I think you can get problems if (a) you're downloading blocks and your power goes out before the data gets synced; I'd expect you might end up with something like "...[marker][nSize=1MB][250kB of block data][64kB of zeroes]" in that case; at which point when you restart you might start writing the next block straight after the zeroes. When running reindex currently, I think your parsing would fail during the 64kB of zeroes, and you'd then scan to the next marker and continue. If so, I guess that means that with this patch, if you ran `-reindex` you'd not see that new block and have to redownload it. That's all speculation though, so don't take my word for it.

Hmm, as an alternative approach, what if you just still read/parse the entire block off the disk, but if it's not one you can immediately accept, you store it into a map for later acceptance? Limit the map to 32MB worth of data, or maybe somehow tie it in to the mempool limit? Keeps the same logic, but reduces the amount of double-reading and double-parsing needed.

Reply:
I did an experiment yesterday in which I simulated corruption in one of the `blk??.dat` files, and as I suspected, all the blocks beyond the coorrupted block were never accepted (because their parents were missing). But the node then seemed to get all the missed blocks from peers, and at the end all seemed good.

Anthony, your suggestion is similar to what this PR originally did before I reduced its scope. You can see the discussion above, but to summarize here, the idea was to increase the size of the circular memory buffer (`CBufferedFile` object) so it could act as a cache, so that we could sometimes access blocks read earlier without having to re-read from disk; the concern was about increasing memory requirements. I think something along the lines I proposed or what you're proposing is likely to help further (especially tying it to mempool limit; that's a good idea), but I suggest keeping this PR limited (no change in memory, a definite improvement) if you don't mind.

Reply:
On the topic of the size being corrupted, if the size field is set to a uniformly distributed random value (or zero, probably the most likely), that is caught (see the check `nSize < 80 || nSize > MAX_BLOCK_SERIALIZED_SIZE` above); the probability that the corrupted value will be in that range is less than 0.1%. So most data corruption there will be detected.